include/pybind11/pybind11.h

```diff
@@ -827,18 +827,25 @@ class module : public object {
     PYBIND11_OBJECT_DEFAULT(module, object, PyModule_Check)
 
     /// Create a new top-level Python module with the given name and docstring
-    explicit module(const char *name, const char *doc = nullptr) {
-        if (!options::show_user_defined_docstrings()) doc = nullptr;
 #if PY_MAJOR_VERSION >= 3
-        PyModuleDef *def = new PyModuleDef();
-        std::memset(def, 0, sizeof(PyModuleDef));
-        def->m_name = name;
-        def->m_doc = doc;
-        def->m_size = -1;
-        Py_INCREF(def);
+    explicit module(const char *name, const char *doc = nullptr, PyModuleDef *def = nullptr) {
+        if (!def) def = new PyModuleDef();
+        def = new (def) PyModuleDef {  // Placement new (not an allocation).
```

Comment:
What happens when a module is reinitialized? I'm not sure this is supported by pybind11 anyway, but I'd like to make sure things don't come crashing down.

Reply:
> What happens when a module is reinitialized? I'm not sure this is supported by pybind11 anyway, but I'd like to make sure things don't come crashing down.

This PR should be as safe as it gets because it follows the recommendation in the Python documentation.

Reinitializing extensions isn't even a supported Python feature IIUC:
https://docs.python.org/3/c-api/init.html#c.Py_FinalizeEx

Reply:
Where do the Python docs talk about overwriting/reusing an existing `PyModuleDef` struct, in case `pybind11_init_...` would be called a second time?

> Reinitializing extensions isn't even a supported Python feature IIUC:
> https://docs.python.org/3/c-api/init.html#c.Py_FinalizeEx

I know; I said I didn't know if it was supported by pybind11. But it would still be nice if things don't explode when a user does try to do this. We've had lots of issues and questions about users trying this before.

Reply:
> But it would still be nice if things don't explode when a user does try to do this.

Great if you have the time to dig into every little eventuality, even if it's almost certain that the time invested will not bear fruits. To me it seems obvious that the best advice to anyone is to not try it, pointing them to the list of Py_FinalizeEx caveats.

This PR is extensively tested, I don't see any reason to dig deeper into theoretical issues. My suggestion is to merge this now. The fact that this is already the 3rd attempt is evidence that it solves a real problem. Then you could use that as a starting point for further refinement in your own time, while the rest of the world is already in a better state compared to right now.

Reply:
> The fact that this is already the 3rd attempt is evidence that it solves a real problem.

1. This PR ignored the 2nd attempt.
2. The fact that this is the 3rd attempt and that the first one was rolled back also means it's a tricky one to get right.

> into theoretical issues

If you don't believe me, ask @bstaletic how many Gitter questions and issues we've already had with people trying to start multiple interpreters and e.g. failing to import numpy as second time.
Things are a lot easier to debug if they don't just print "Segmentation fault.", and all I'm asking is if it doesn't.

Reply:
> If you don't believe me, ask @bstaletic how many Gitter questions and issues we've already had with people trying to restart the interpreter and e.g. failing to import numpy as second time.

Is there a situation where that actually works reliably?

Reply:
Using pybind11 to embed Python and importing pure Python libraries (and some simple extension libraries), yes, I believe. Reloading pybind11 modules, almost never, I think, because `class_` will complain about types already being registered. And e.g. numpy is also too complex for this to work. But I don't believe things segfault, currently (or do they, @bstaletic?).

But once every so often, we do get an issue or question, so I just wanted to know whether this will cause segfaults. It's a C struct, so one would hope `memset` is fine, but...

Reply:
> But once every so often, we do get an issue or question, so I just wanted to know whether this will cause segfaults. It's a C struct, so one would hope `memset` is fine, but...

My estimate: holding back this fix has a cost that's far higher than maybe (!) dealing with questions around a situation that's practically certain to be invalid anyway. You have so many "believe" and "think" in what you wrote, there isn't even a meaningful starting point for trying to tease out a segfault.

Reply:
> holding back this fix has a cost that's far higher than maybe (!) dealing with questions around a situation

What's the high cost of holding back this fix a little while longer than the 13 days it's old?
You're not the one answering questions, most of the time, either.

I don't think this is a blocking issue, though, but that doesn't mean the concern can just be dismissed.

> You have so many "believe" and "think" in what you wrote, there isn't even a meaningful starting point for trying to tease out a segfault.

Yes, trying to be careful not to overstate or misremember. UB and segfault are always tricky, as you certainly know. Doesn't invalidate the question I'm asking, though.

Reply:
> What's the high cost of holding back this fix a little while longer than the 13 days it's old?

N people out there continuing to stumble over this, while this PR is not merged over fringe concerns that can easily be addressed in isolation, in follow-up PRs.

Reply:
For which value of N, and which definition of stumble? This memory leak has been there from the start, similar like all the others in CPython's valgrind suppression file.